feat(AIP-111): introducing AIP around planes.

toumorokoshi opened the pull request:
The term data plane have been used in the AIPs without a clear
definition.

Introducing the term "management" and "data" plane, as these are
industry-standard terms and are also well defined in the context of
Google APIs.

These definitions are also helpful in:

- providing fine-grained API guidance depending on the plane
- help inform new annotations on resources and methods, which can be
  used for auditing and determining coverage
- further help clarify classifications of methods (AIP-130)

Also removing the redundant text in AIP-128.

toumorokoshi committed:
WIP defining planes of operations

toumorokoshi committed:
feat(AIP-111): introducing plane AIP

The term data plane have been used in the AIPs without a clear
definition.

Introducing the term "management" and "data" plane, as these are
industry-standard terms and are also well defined in the context of
Google APIs.

These definitions are also helpful in:

- providing fine-grained API guidance depending on the plane
- help inform new annotations on resources and methods, which can be
  used for auditing and determining coverage
- further help clarify classifications of methods (AIP-130)

Also removing the redundant text in AIP-128.

toumorokoshi committed:
Merge branch 'master' into toum/plane-explanation

toumorokoshi committed:
fixing build

- replacing 0008 front matter with correct example
- fixing 111 front matter.

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+or perform operations upon. For the context of APIs, the following planes are
+defined:
+
+- The management plane: a uniform, resource-oriented API that primarily
```
We probably need to at least mention the control plane here, even if it may not be distinct from the management plane, since that term is likely more familiar. 

Also, some people may consider the IaC layer the management plane and the resource APIs to be the control plane.

toumorokoshi replied:
That begs the question: do we want to introduce a term like the management plane? It's a bit easier to define operations in terms of their impact on the control plane vs data-plane.

Although I guess this wouldn't capture that nuance that sometimes an operation may want to operate on both (e.g. control plane provisions a database and immediately populates it with a table or permissions).

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+The following behaviors are common (but not required) attributes of a
+management plane resource:
+
+- changes infrequently.
```
This won't be true in the case of control loops, such as autoscaling, built on the standard resource APIs.

noahdietz replied:
What does this even mean though? An instance of the resource doesn't change often? The definition of the resource doesn't change often? It's a tad vague imo

bgrant0607 replied:
We need consistent terminology for referring to (a) the resource attributes, sometimes known as metadata, but that's ambiguous since fields like name, creation time, and annotations are also referred to as metadata, and (b) user data.

toumorokoshi replied:
for now I can remove this section - we can always add something to that effect later, and I think the "major distinctions" section does a good job highlighting key differences.

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+
+- The management plane: a uniform, resource-oriented API that primarily
+  configures and allows retrieval of resources.
+- The data plane: a heteregenous API (ideally resource-oriented) that interface
```
This is missing the key aspect of the data plane, which is that it writes and reads user data. In the case of resource-oriented APIs, the data are associated with resources.

bgrant0607 replied:
We may also want to include "and/or connects to provisioned entities, such as load balancers or virtual machines"

toumorokoshi replied:
rephrased to emphasize reading / writing of user data, connected to entities provisioned by the management plane. PTAL.

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+
+- virtual machines
+- virtual private networks
+- disks
```
Virtual disks (for consistency)

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+
+### Data Plane
+
+Methods on the data plane operate on a variety of data formats, and generally
```
Methods on the data plane operate on user data in a variety of formats.

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+interface with a resource provisioned via a management plane API. Examples of
+data plane methods include:
+
+- tables in a database
```
"tables" are not methods, which are verbs. They are nouns. Anyway, I wouldn't pick ambiguous examples. This is more about 3P vs 1P. Kubernetes is all control plane / management plane, but hosted 3P/OSS.

toumorokoshi replied:
Sure, I'd like to round back on these I can remove the example.

For 3P/OSS I'll follow up in the other thread.

toumorokoshi replied:
Also ack on tables not being methods. I guess we are looking at method categorization here.

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+data plane methods include:
+
+- tables in a database
+- rows in a table
```
writing and querying/reading rows in a table

toumorokoshi replied:
I'll leave it as writing or reading rows, since querying is implied. feel free to reopen if you think we need the querying part.

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+
+- tables in a database
+- rows in a table
+- pushing or pulling from a message queue
```
pushing to

bgrant0607 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+context. Therefore, the definition of management or data plane is incomplete
+without stating the context explicitly.
+
+For example, consider a cloud provider which exposes a Kubernetes cluster
```
I suppose this gets at the 1P vs hosted 3P issue. I would frame it more in those terms.

toumorokoshi replied:
It it always 1P vs 3P? for example, BigQuery is a 1P service, but certainly operations against the bigquery table are data-plane? 

I do agree that 1P vs 3P provides a clean delineation where applicable. That could be the first question to determine whether your operation is in the management or data plane.

bgrant0607 replied:
Tables are resources: https://cloud.google.com/bigquery/docs/reference/rest/v2/tables

CRUDL operations on tables would be management plane.

If they were defined with SQL statements, that would be 3P.

Reading/writing data in the table is data plane.

toumorokoshi replied:
I think I'm going to remove this section - to your point it speaks more to 1st and third-party APIs, and we need a separate section on that one when defining the scope of the AIPs. 

There's already a section that talks about data-plane operations being exposed in a resource-oriented fashion.

alin04 commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+
+- For convenience, resources and methods that operate on the data plane **may**
+expose themselves via resource-oriented management APIs. If so, those resources
+and methods **must** adhere to the requirements of the management plane.
```
What are `requirements of the management plane`?

toumorokoshi replied:
one is in AIP-133 around user-settable IDs management APIs should allow user-settable IDs on resources while data-plane does not.

I'd rather not enumerate them here because we have issues with maintaining duplicate guidance in multiple AIPs: https://github.com/aip-dev/google.aip.dev/pull/1139.

I added a sentence to note that the guidance is in other AIPs. WDYT?

bgrant0607 replied:
I think that refers to the resource-oriented AIPs. It would be useful to refer to that explicitly.

toumorokoshi replied:
added!

noahdietz commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+or perform operations upon. For the context of APIs, the following planes are
+defined:
+
+- The management plane: a uniform, resource-oriented API that primarily
```
Nit:
```suggestion
- Management plane: a uniform, resource-oriented API that primarily
```

noahdietz commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+
+- The management plane: a uniform, resource-oriented API that primarily
+  configures and allows retrieval of resources.
+- The data plane: a heteregenous API (ideally resource-oriented) that interface
```
Nit:
```suggestion
- Data plane: a heterogenous API (ideally resource-oriented) that interface
```

noahdietz commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+The following behaviors are common (but not required) attributes of a
+management plane resource:
+
+- changes infrequently.
```
Nit:
```suggestion
- changes infrequently
```

noahdietz commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+- tables in a database
+- rows in a table
+- pushing or pulling from a message queue
+- uploading or downloading blobs from a blob store instance
```
uploading blobs to

noahdietz commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+### Major distinctions between management and data plane
+
+- [IaC][] tools operate on the management plane exclusively.
+- Data planes provide much higher reliability, since they are often critical to
```
Data planes are much more performance sensitive than management planes.

Not sure if the broader "reliability" term was meant to address that.

bgrant0607 replied:
I interpreted it as availability

noahdietz replied:
Yeah, so did I. I think performance/high QPS capabilities are an important difference (we had a good discussion in our morning meeting about this, actually). Might be worth mentioning, but not blocking if it needs more discussion.

toumorokoshi replied:
Yes, availability might be a better term. I've added the points on both performance-sensitive and higher-throughput requirements.

noahdietz commented on aip/general/0111.md:
```diff
@@ -0,0 +1,90 @@
+For example, consider a cloud provider which exposes a Kubernetes cluster
+resource, which in turn exposes it's own API with resources. In the context of
+the cloud, the resources which live in the Kubernetes cluster are *data plane*.
+In the context of the Kubernetes cluster, the resources are *management plane*.
```
When you say "resources" in here do you mean methods? b.c AIP-121 resources aren't "data plane". In some of the "management plane resources" above, they are all nouns, not verbs. So I just want a little clarity here.

bgrant0607 replied:
+1. This is about hosted 3P/OSS.

toumorokoshi replied:
sorry about that, I uploaded the change to make it about methods.

toumorokoshi replied:
removed the planes are contextual section entirely, will replace with a "1p vs 3p" discussion in the AIPs in a separate PR.

noahdietz commented on aip/general/0128.md:
```diff
@@ -26,20 +26,6 @@ tools requires uniformity, so that integration can be automated.
 
 ## Guidance
 
-Services **should** clearly delineate between "control plane" operations and
```
Add a changelog entry please.

noahdietz commented on aip/general/0133.md:
```diff
@@ -126,8 +126,16 @@ to done if the request is effectively immediate.
-An API **must** allow a client to specify the ID component
-of a resource (the last segment of the resource name) on creation. An API
+An API **must** allow a user to specify the ID component
+of a resource (the last segment of the resource name) on creation if the API is operating on the management plane.
```
Nit:

```suggestion
of a resource (the last segment of the resource name) on creation if the API is operating on the [Management Plane](./0111.md#management-plane).
```

noahdietz commented on aip/general/0133.md:
```diff
@@ -126,8 +126,16 @@ to done if the request is effectively immediate.
+An API **must** allow a user to specify the ID component
+of a resource (the last segment of the resource name) on creation if the API is operating on the management plane.
+
+On the data plane, an API **should** allow a user to specify the ID. Exceptional cases should have the following behavior:
```
Nit:

```suggestion
On the [Data Plane](./0111.md#data-plane), an API **should** allow a user to specify the ID. Exceptional cases should have the following behavior:
```

toumorokoshi committed:
Addressing feedback

- adding changelog entries and proper links.
- emphasizing data-plane's role in reading and writing user data.
- accepting suggestion and nits.

noahdietz commented on aip/general/0111.md:
```diff
@@ -0,0 +1,93 @@
+---
+id: 111
+state: draft
```
When can we change it to approved?...before submitting? :D

Love this guidance

noahdietz commented on aip/general/0008.md:
```diff
@@ -38,10 +38,9 @@ AIPs **must** have appropriate front matter.
-  state: reviewing
-  created: 2019-05-28
+id: 8
+state: reviewing
```
I think this should be approved? The site has it as approved anyways...weird
![image](https://github.com/aip-dev/google.aip.dev/assets/6644735/015b02d2-ae5a-4a7d-aa8c-3f9686b49034)

toumorokoshi replied:
This is the template, not the actual site. 

this was supposed to be a separate fix, but the frontmatter template was broken.

noahdietz reviewed (approved):
I'll be the first stamp I guess, pending resolution of other clarity comments

toumorokoshi committed:
addressing more feedback.

bgrant0607 commented on aip/general/0133.md:
```diff
@@ -126,8 +126,18 @@ to done if the request is effectively immediate.
 
-An API **must** allow a client to specify the ID component
-of a resource (the last segment of the resource name) on creation. An API
+An API **must** allow a user to specify the ID component of a resource (the last
```
Any more updates planned to this section, such as the global namespace issue, whether renaming is needed, whether the resource is created as a side effect of a custom method rather than the standard Create method, etc.?

toumorokoshi replied:
not for this PR - I plan on sending another one. Does that sound good?

bgrant0607 replied:
Yes

toumorokoshi committed:
Merge branch 'master' into toum/plane-explanation

toumorokoshi committed:
Addressing comments

- removing planes are contextual.

bgrant0607 reviewed (approved):
Thanks. LGTM

toumorokoshi committed:
fix typo
